db/compatibility/compatibility_db_contents.go

```diff
@@ -0,0 +1,114 @@
+package compatibility
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/agglayer/aggkit/common"
+	"github.com/agglayer/aggkit/db"
+	comptypes "github.com/agglayer/aggkit/db/compatibility/types"
+	"github.com/agglayer/aggkit/log"
+)
+
+/*
+This file contains the compatibility check logic:
+The usage is:
+- Add a field `CompatibilityChecker` in your class
+- Create a checker with the NewCompatibilityCheck function
+- Invoke the Check method in your class
+*/
+
+// CompatibilityChecker is the interface that defines the methods to check the compatibility
+// the object CompatibilityCheck[T] implements this interface
+type CompatibilityChecker interface {
+	Check(ctx context.Context, tx db.Querier) error
+}
+
+// CompatibilityDataStorager is the interface that defines the methods to interact with the storage
+type CompatibilityDataStorager[T any] interface {
+	// GetCompatibilityData returns the compatibility data from the storage:
+	// true -> if data is stored / false -> if data is not stored yet
+	// T -> the data stored
+	// error -> if there is an error
+	GetCompatibilityData(ctx context.Context, tx db.Querier) (bool, T, error)
+	// SetCompatibilityData stores the compatibility data in the storage
+	// error -> if there is an error
+	SetCompatibilityData(ctx context.Context, tx db.Querier, data T) error
+}
+
+// RuntimeDataGetterFunc is a function that returns the runtime data
+type RuntimeDataGetterFunc[T comptypes.CompatibilityComparer[T]] func(ctx context.Context) (T, error)
+
+// CompatibilityCheck is the object that checks the compatibility between the runtime data and
+// the data stored in the storage. It's the implementation of the CompatibilityChecker interface
+type CompatibilityCheck[T comptypes.CompatibilityComparer[T]] struct {
+	RequireStorageContentCompatibility bool
+	RuntimeDataGetter                  RuntimeDataGetterFunc[T]
+	Storage                            CompatibilityDataStorager[T]
+	Logger                             common.Logger
+}
+
+// NewCompatibilityCheck creates a new CompatibilityCheck object
+func NewCompatibilityCheck[T comptypes.CompatibilityComparer[T]](
+	requireStorageContentCompatibility bool,
+	runtimeDataGetter RuntimeDataGetterFunc[T],
+	storage CompatibilityDataStorager[T]) *CompatibilityCheck[T] {
+	return &CompatibilityCheck[T]{
+		RequireStorageContentCompatibility: requireStorageContentCompatibility,
+		RuntimeDataGetter:                  runtimeDataGetter,
+		Storage:                            storage,
+	}
+}
+
+// Check checks the compatibility between the runtime data and the data stored in the storage
+func (s *CompatibilityCheck[T]) Check(ctx context.Context, tx db.Querier) error {
+	err := s.initialize()
+	if err != nil {
```

Comment:
And on other places to lower affected lines number:

```suggestion
	if err := s.initialize(); err != nil {
```